api/agent/lb_agent.go

```diff
@@ -0,0 +1,182 @@
+package agent
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/fnproject/fn/api/models"
+	"github.com/fnproject/fn/fnext"
+	"github.com/fnproject/fn/poolmanager"
+)
+
+// RequestReader takes an agent.Call and return a ReadCloser for the request body inside it
+func RequestReader(c *Call) (io.ReadCloser, error) {
+	// Get the call :(((((
+	cc, ok := (*c).(*call)
+
+	if !ok {
+		return nil, errors.New("Can't cast agent.Call to agent.call")
+	}
+
+	if cc.req == nil {
+		return nil, errors.New("Call doesn't contain a request")
+	}
+
+	logrus.Info(cc.req)
+
+	return cc.req.Body, nil
+}
+
+func ResponseWriter(c *Call) (*http.ResponseWriter, error) {
+	cc, ok := (*c).(*call)
+
+	if !ok {
+		return nil, errors.New("Can't cast agent.Call to agent.call")
+	}
+
+	if rw, ok := cc.w.(http.ResponseWriter); ok {
+		return &rw, nil
+	}
+
+	return nil, errors.New("Unable to get HTTP response writer from the call")
+}
+
+// The LB agent performs its functionality by delegating to a remote node. It
+// pretends to have allocated a slot, and slot.exec() is what actually handles
+// the protocol with the remote node; this Slot implementation is used.
+type remoteSlot struct {
+	lb *lbAgent
+}
+
+func (s *remoteSlot) exec(ctx context.Context, call *call) error {
+	// TODO: do it properly!
+	a := s.lb
+
+	memMb := call.Model().Memory
+	lbGroupID := GetGroupID(call.Model())
+
+	capacityRequest := &poolmanager.CapacityRequest{TotalMemoryMb: memMb, LBGroupID: lbGroupID}
+	a.np.AssignCapacity(capacityRequest)
+	defer a.np.ReleaseCapacity(capacityRequest)
+
+	err := a.placer.PlaceCall(a.np, ctx, call, lbGroupID)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to place call")
+	}
+	return err
+}
+
+type Placer interface {
+	PlaceCall(np NodePool, ctx context.Context, call *call, lbGroupID string) error
+}
+
+type naivePlacer struct {
+}
+
+func NewNaivePlacer() Placer {
+	return &naivePlacer{}
+}
+
+func (sp *naivePlacer) PlaceCall(np NodePool, ctx context.Context, call *call, lbGroupID string) error {
+	deadline := call.slotDeadline
+
+	for {
+		if time.Now().After(deadline) {
+			return models.ErrCallTimeoutServerBusy
+		}
+
+		for _, r := range np.Runners(lbGroupID) {
+			placed, err := r.TryExec(ctx, call)
+			if err != nil {
+				logrus.WithError(err).Error("Failed during call placement")
+			}
+			if placed {
+				return err
+			}
+
+		}
+
+		time.Sleep(retryWaitInterval)
```

Comment:
same issue as lb_hash side, we should select on both ctx and retryWaitInterval to handle ctx timing out or ctx being cancelled.